pytensor/xtensor/shape.py

```diff
@@ -73,6 +75,135 @@ def stack(x, dim: dict[str, Sequence[str]] | None = None, **dims: Sequence[str])
     return y
 
 
+def expand_ellipsis(
+    dims: tuple[str, ...], all_dims: tuple[str, ...], validate: bool = True
+) -> tuple[str, ...]:
+    """Expand ellipsis in dimension permutation.
+
+    Parameters
+    ----------
+    dims : tuple[str, ...]
+        The dimension permutation, which may contain ellipsis
+    all_dims : tuple[str, ...]
+        All available dimensions
+    validate : bool, default True
+        Whether to check that all non-ellipsis elements in dims are valid dimension names.
+
+    Returns
+    -------
+    tuple[str, ...]
+        The expanded dimension permutation
+
+    Raises
+    ------
+    ValueError
+        If more than one ellipsis is present in dims.
+        If any non-ellipsis element in dims is not a valid dimension name and validate is True.
+    """
+    if dims == () or dims == (...,):
+        return tuple(reversed(all_dims))
+
+    if ... not in dims:
+        if validate:
+            invalid_dims = set(dims) - set(all_dims)
+            if invalid_dims:
+                raise ValueError(
+                    f"Invalid dimensions: {invalid_dims}. Available dimensions: {all_dims}"
+                )
+        return dims
+
+    if sum(d is ... for d in dims) > 1:
+        raise ValueError("an index can only have a single ellipsis ('...')")
+
+    pre = []
+    post = []
+    found = False
+    for d in dims:
+        if d is ...:
+            found = True
+        elif not found:
+            pre.append(d)
+        else:
+            post.append(d)
+    if validate:
+        invalid_dims = set(pre + post) - set(all_dims)
+        if invalid_dims:
+            raise ValueError(
+                f"Invalid dimensions: {invalid_dims}. Available dimensions: {all_dims}"
+            )
+    middle = [d for d in all_dims if d not in pre + post]
+    return tuple(pre + middle + post)
+
+
+class Transpose(XOp):
+    __props__ = ("dims", "missing_dims")
+
+    def __init__(
+        self,
+        dims: tuple[str, ...],
```

Comment:
Type-hint is still wrong?

```suggestion
        dims: tuple[str | Literal[...], ...],
```

Or I guess you can use `EllipsisType`. `tuple[str, ...]` means a tuple of variable length with string elements only, but you are allowing string or ellipsis (of varying length)